lib/ansible/module_utils/basic.py

```diff
@@ -1905,6 +1909,12 @@ def _set_cwd(self):
         # and we don't want to break modules unnecessarily
         return None
 
+    # TODO: move this to module method, only self.fail_json is used from self, and
+    #       that could/should just raise an exception, or second choice would be to return
+    #       a results data structure (bin_path, errors) for simplest case.
+    #       Why? because get_bin_path() is one of the few methods
+    #       facts use from AnsibleModule(). If module methods they wouldn't require an
+    #       instance of AnsibleModule
```

Comment:
This is a good idea, however, I'd say it should apply to most of what's in AnsibleModule.  We probably want to make these notes in a discussion about the next version of the standard module API.  Probably list out all of the functions and methods in basic.py and do an analysis there of each of them, trying to organize and break them apart appropriately.  (and how to have them recombined in basic.py to provide a compatible API at the same time).

Also note that basic.py is included with every module (thus being downloaded to every machine for every task) so keeping future notes in the file is not a good idea.